src/Type.h

```diff
@@ -90,25 +93,25 @@ struct halide_handle_cplusplus_type {
 
     /// Qualifiers and indirections on type. Allows up to 8 levels. 0 is innermost.
     struct CPPTypeModifiers {
-        uint8_t data[8];
-      CPPTypeModifiers(const std::vector<uint8_t> &vals) {
-            user_assert(vals.size() <= sizeof(data)) << "Too many levels of indirection in handle type " << vals.size() << " where " << sizeof(data) << " are allowed\n";
-            std::copy(vals.begin(), vals.end(), &data[0]);
-            std::fill(&data[vals.size()], &data[sizeof(data)], 0);
+        std::array<uint8_t, 8> data;
+        CPPTypeModifiers(const std::vector<uint8_t> &vals) {
+            user_assert(vals.size() <= data.size()) << "Too many levels of indirection in handle type " << vals.size() << " where " << sizeof(data) << " are allowed\n";
+            std::copy(vals.begin(), vals.end(), data.begin());
+            std::fill(data.begin() + vals.size(), data.end(), 0);
         }
         const uint8_t &operator[](size_t index) const { return data[index]; }
         bool operator==(const CPPTypeModifiers &rhs) const {
-            return std::equal(&data[0], &data[sizeof(data)], &rhs.data[0]);
+            return std::equal(data.begin(), data.end(), rhs.data.begin());
         }
         const uint8_t *begin() const { return &data[0]; }
-        const uint8_t *end() const { return &data[sizeof(data)]; }
+        const uint8_t *end() const { return begin() + data.size(); }
```

Comment:
Please change this to "&data[0] + data.size()" to make it clear what is going on.

IIRC, the general reason I bailed on std::array is because it is impossible to construct the end pointer while staying inside the std::array interface. I.e. all this code here placates Microsoft's compiler warnings while offering no actual protection because one has to convert to pointers anyway.